linkerd/cache/src/lib.rs

```diff
@@ -19,32 +21,117 @@ where
 {
     inner: N,
     services: Services<T, S>,
+
+    // As long as this is held, the eviction tasks will activate every time it is
+    // notified.
+    evict: Arc<Notify>,
 }
 
 /// A tracker inserted into each inner service that, when dropped, indicates the
 /// service may be removed from the cache.
-pub type Handle = Arc<()>;
+#[derive(Debug)]
+pub struct Handle {
+    inner: Arc<()>,
+    evict: Weak<Notify>,
+}
 
 type Services<T, S> = Arc<RwLock<HashMap<T, (S, Weak<()>)>>>;
 
+// === impl Handle ===
+
+impl Drop for Handle {
+    fn drop(&mut self) {
+        if let Some(evict) = self.evict.upgrade() {
+            evict.notify_one();
+        }
+    }
+}
+
 // === impl Cache ===
 
 impl<T, N> Cache<T, N, N::Service>
 where
-    T: Eq + Hash,
-    N: NewService<(Handle, T)>,
+    T: Eq + Hash + Send + Sync + 'static,
+    N: NewService<(Handle, T)> + 'static,
+    N::Service: Send + Sync + 'static,
 {
     pub fn layer() -> impl layer::Layer<N, Service = Self> + Copy + Clone {
-        layer::mk(|inner| Self {
-            inner,
-            services: Services::default(),
+        layer::mk(|inner| {
+            let (cache, _) = Self::spawn(inner);
+            cache
         })
     }
 
-    fn new_entry(new: &mut N, target: T) -> (N::Service, Weak<()>) {
-        let handle = Arc::new(());
-        let weak = Arc::downgrade(&handle);
-        let svc = new.new_service((handle, target));
+    fn spawn(inner: N) -> (Self, tokio::task::JoinHandle<()>) {
+        let services = Services::default();
+        let evict = Arc::new(Notify::new());
+
+        // Spawn a background task to remove defunct services.
+        let task = tokio::spawn(
+            Self::evict(Arc::downgrade(&evict), services.clone()).instrument(debug_span!("evict")),
+        );
+
+        let cache = Self {
+            inner,
+            services,
+            evict,
+        };
+        (cache, task)
+    }
+
+    // Evicts services every time the evict handle is notified, as long as the
+    // evict handle remains active.
+    async fn evict(evict: Weak<Notify>, services: Services<T, N::Service>) {
+        loop {
+            // If the cache still holds the eviction handle, wait for it to be
+            // notified before evicting services. It is notified whenever a
+            // handle is dropped and when the cache is dropped.
+            match evict.upgrade() {
+                Some(e) => {
+                    trace!("Awaiting eviction signal");
+                    e.notified().await;
+                    trace!("Evicting services");
+                }
+                None => {
+                    debug!("Cache dropped");
+                    return;
+                }
+            }
+
+            // Drop defunct services before inserting the new service into the
+            // cache.
```

Comment:
Nit: this comment looks like it was copied from a previous impl where eviction happened when inserting a new service.  Now, the eviction is performed in the background, so we might want to update this:
```suggestion
            // Drop defunct services.
```